Add support for running with Python 3.6 and 3.7

Eric-Arellano opened the pull request:
### Problem
Now that we're releasing Pants with Python 3 wheels (https://github.com/pantsbuild/pants/pull/7197, overall migration tracked by https://github.com/pantsbuild/pants/issues/6062), we need to add a way for every day Pants users to consume the upgrade.

See https://github.com/pantsbuild/setup/issues/30 for more context, including the design goals.

### Solution
Parse the `pants_runtime_python_version` from `pants.ini` to determine which Python interpreter to use. If it's missing, default to Python 2.7.

This requires modifying the venv folders to support multiple Python versions. Rather than using the naming scheme `${pants_version}`, we now consistently use `${pants_version}_py{py_major_minor}`, regardless of if the user is using `pants_runtime_python_version`.

It also requires upgrading virtualenv to avoid a deprecation warning when running with Python 3.

Finally, we add tests to ensure this all works.

### Result
Running `./pants` like normal for the first time will bootstrap Pants again and create the new folder `1.15.dev3_py{27,36,37}`.

Changing the `$PYTHON` env var or the `pants_runtime_python_version` in `pants.ini` will change which Python version is used to run Pants.

When changing interpreters for the first time, it will bootstrap Pants again. After that, you can change without issue.

Eric-Arellano committed:
Parse targeted Python version via pants.ini value

Rather than using a sentinal value, read in the pants.ini value for `pants_engine_python_version`. Default to Py27 because that's what we've always used and most people will have this.

Eric-Arellano committed:
Update docs to mention pinning Python version

Eric-Arellano committed:
Remove pants_engine_python_version from pants.ini

We first need to land https://github.com/pantsbuild/pants/pull/7363 and then release Pants with it or else the option will error out.

Eric-Arellano committed:
Fix interpreter_version failing script when interpreter_version not found in pants.ini

For some reason it was exiting the script?! I have no idea why, because the same code works for parsing the pants_version.

This does the same thing and has the benefit of using way less lines.

Eric-Arellano committed:
Add ci.py to CI to allow parameterizing pants_version and python_version

Eric-Arellano committed:
Squashed commit of the following:

commit e8cf735e23db657e6ba26825024bc063825a472a
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 09:36:16 2019 -0700

    Trusty Py34 is from system, not pyenv

commit 1f7d23685a7cdc127c735dfcbd727f4a494bacd4
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 09:30:08 2019 -0700

    OSX requires language: general

commit 4941956ca20518c63d40d29f5ec3b66604b22a71
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 09:23:28 2019 -0700

    Move constant to top of file

    Think this is the style used in most Python projects and ours.

commit c478ecb64e3eaff9cbecd144375996551474af8b
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 09:19:29 2019 -0700

    Set Pyenv for Trusty

commit 30aab56793cf45dec9b4c21951c9c084b5de5dee
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 09:16:58 2019 -0700

    Try pyenv approach

    Path wrangling didn't work.

commit 9b9e6be85b19d590ef033cf580724b7bd9c4e928
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 09:12:11 2019 -0700

    Global python entry -> build matrix

    See if moving back to osx_setup and linux_setup works.

commit 0dc87dcb99740fd6aa14724282de2c888deaf49e
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 09:09:18 2019 -0700

    Maybe fix Trusty using Py3.4

commit aaab4bf112ad03310202224bdb89236ee2aff927
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 09:05:50 2019 -0700

    Test if OSX comes pre-installed with Python?

commit 96cb6ac412dd7427c78722f1630bc30879e600e8
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 09:02:44 2019 -0700

    Remove unneeded anchors from OSX shards

    Turns out Precise comes installed with 3.6 when requested! So we only need Pyenv for OSX, not Linux.

commit 4a312a42ae61b5e5e6341efd1af5631103b45d52
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 09:01:31 2019 -0700

    Try to debug Trusty image using Py3.4

commit d511d99ef083582b5b5619a3374c723619d6510d
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 08:48:28 2019 -0700

    Remove stray whitespace

commit 126ae2f784025053be7ae13f7fc57a9d54d4324f
Merge: cd28677 4aee51a
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 08:47:56 2019 -0700

    Merge branch 'gh-pages' of github.com:pantsbuild/setup into ci-py

commit 4aee51ad08ad2778fca0a0ce497ac3ef3a71ed5b
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 08:46:38 2019 -0700

    Test Ubuntu Xenial (#31)

    Travis added Xenial back in November. See https://blog.travis-ci.com/2018-11-08-xenial-release.

    Adding this test results in better coverage.

commit cd286772472f081a28de7664b36c7319fe713424
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 08:36:55 2019 -0700

    Add  prefix to script. Oops

commit 4afcdcfd1921af4341b55cc1d0b04b0f00ecb6b7
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 08:24:43 2019 -0700

    Add comment explaining why we rewrite pants.ini like this

commit ef5bf71ad99020dcb499476a1569e86aa5460a5a
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 08:22:55 2019 -0700

    Tweak ci.py formatting

commit 599ae46640b4c0f520fa8fcdc9f60dca5245a695
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 08:19:38 2019 -0700

    Try to install Py36 on each shard

    Also modernize .travis.yml to use all the cool goodies we've learned in Pants's CI like anchors and the brew addon package!

commit 717271fbb9cba68a9ff74d8f112b23661860198a
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 07:50:10 2019 -0700

    Remove PythonVersion and fix --pants-version unspecified

    --pants-version unspecified would fail because of a bad config, that the contrib module would refer to pants_version even though that line was deleted.

commit d1f9f0f1b3065f507cf05817c45fe5e0429f7629
Author: Eric Arellano <ericarellano@me.com>
Date:   Tue Mar 12 07:40:38 2019 -0700

    Add ci.py to CI to allow parameterizing pants_version and python_version

Eric-Arellano committed:
Implement naming scheme

Will look like this: 1.14.dev1_py36.

Eric-Arellano committed:
Update interpreter selection test to modify pants.ini

We won't be able to resolve Py3 from PyPi until the release anyway, so we should setup the test correctly from the start. Once the Pypi release happens, we can rebase and get green CI

Eric-Arellano committed:
Rename venv folders to py27 etc., not py2.7

Easier on the eyes.

Eric-Arellano committed:
Upgrade version to 1.15.0.dev4

We'll need this version to get CI passing so that it can consume Py3 wheels and parse pants_engine_python_version.

Eric-Arellano committed:
Merge branch 'gh-pages' of github.com:pantsbuild/setup into py3

Eric-Arellano commented:
Reviewers: CI is going to fail on this one until we release `1.15.0.dev4`, because we need the Py3 wheels and https://github.com/pantsbuild/pants/pull/7363 to parse `pants_engine_python_version` in `pants.ini`.

Stu and I came up with a strategy to (hopefully) still land this change before sending the release email to pants-devel. Right after we finish the release to PyPi, I'll update the `pants.ini` to point to the new version and run CI. If all goes green, we'll merge this PR and send the email to pants-devel about how to consume this upgrade. If CI fails, we simply push off the announced release one week and go with a silent release.

stuhood commented on ci.py:
```diff
@@ -68,5 +87,37 @@ def setup_pants_version(test_pants_version: PantsVersion):
+@contextmanager
+def setup_python_version(test_python_version: PythonVersion):
+  """Modify pants.ini to allow the Python version to be unspecified or change to what was requested."""
+  expected_prefix = "pants_engine_python_version"
```
This was renamed in https://github.com/pantsbuild/pants/pull/7363 I think. Ditto the PR description, and a few other locations in here.

stuhood commented on ci.py:
```diff
@@ -68,5 +87,37 @@ def setup_pants_version(test_pants_version: PantsVersion):
 
 
+@contextmanager
+def setup_python_version(test_python_version: PythonVersion):
```
This is a pretty large method... might be good to extract some helpers.

Eric-Arellano replied:
Good idea! I refactored it. Let me know if it could use another refactor.

stuhood commented on pants:
```diff
@@ -13,11 +13,15 @@
+  grep -E "^[[:space:]]*pants_engine_python_version" pants.ini 2>/dev/null | \
+    cut -f2 -d: | tr -d " "
+) || interpreter_version=2.7
+PYTHON=${PYTHON:-$(which python${interpreter_version})}
```
Is it worth checking that this resulted in a non-empty value, and erroring if not? Or does the `venv` creation result in a useful error message if your choice isn't installed?

stuhood reviewed (approved):
Thanks. One blocker related to the option name, but the rest looks reasonable.

Eric-Arellano committed:
Rename to pants_runtime_python_version

Eric-Arellano committed:
Refactor ci.py

Eric-Arellano committed:
Merge branch 'gh-pages' of github.com:pantsbuild/setup into py3

Eric-Arellano committed:
Check interpreter is valid

Copy idiom we use in Pants' `virtualenv` file.

Eric-Arellano committed:
Fix not using keyword arg

Eric-Arellano committed:
Remove stry pdb import

Eric-Arellano committed:
Nest conditionals in write_config_entry

Usually I don't like nested conditionals, but it reads better here.

Eric-Arellano committed:
Fix one more bad reference to pants_engine_python_version

Eric-Arellano committed:
No longer change documentation

We shouldn't do this till 1.15.0 is cut. See https://pantsbuild.slack.com/archives/CBNMV1LRH/p1552535884071700.

Eric-Arellano committed:
Skip test that will always fail until 1.15.0

Eric-Arellano committed:
Also remove pants_runtime_python_version from pants.ini

Until 1.15.0, we shouldn't have this here as it's an example new users might refer to, and we don't want them to try putting this value then using a version < 1.15.0.dev4 and being frustrated it doesn't work.

We don't even need it here because our ci.py script dynamically rewrites pants.ini to add this value during tests.

Eric-Arellano committed:
Add back accidently deleted <p>

Eric-Arellano committed:
Improve wording

cosmicexplorer commented on pants:
```diff
@@ -13,11 +13,21 @@
 PANTS_BOOTSTRAP="${PANTS_HOME}/bootstrap-$(uname -s)-$(uname -m)"
 
+interpreter_version=$(
+  grep -E "^[[:space:]]*pants_runtime_python_version" pants.ini 2>/dev/null | \
```
I still definitely think this should:
1. be its own function so you can call it for either `pants_version` or `pants_runtime_python_version`
2. be converted into `sed -rne '/\s*pants_version: / s#^.*: ##p'` as per https://github.com/pantsbuild/setup/issues/30#issuecomment-471805216, as in:
```bash
function get_pants_ini_config_line() {
  config_key="$1"
  sed -rne "/\s*${config_key}: / s#^.*: ##p" pants.ini 2>/dev/null
}
```
If that's too terse, comments from https://github.com/pantsbuild/setup/issues/30#issuecomment-471805216 could be added. I would at least like this to be made into its own function, but using `sed` here gives us more flexibility later if we need it without adding more complexity.

cosmicexplorer commented on pants:
```diff
@@ -77,7 +87,10 @@ function bootstrap_pants {
   fi
 
-  if [[ ! -d "${PANTS_BOOTSTRAP}/${pants_version}" ]]
+  python_major_minor_version=$(${PYTHON} -c 'import sys; print("".join(map(str, sys.version_info[0:2])))')
```
I like the idea of making this into a function so it's less hidden that we're invoking a python subprocess to do something nontrivial here. It also makes it easier to review later. I would also like it if this function used a heredoc, e.g.:
```bash
function python_major_minor_version() {
  python_exe="$1"
  "$python_exe" <<EOF
import sys
major_minor_version = ''.join(str(version_num) for version_num in sys.version_info[0:2])
print(major_minor_version)
EOF
}
```

That's intentionally verbose to highlight how you have more room to work with when making it into a separate function.

cosmicexplorer commented on ci.py:
```diff
@@ -35,37 +46,83 @@ def create_parser() -> argparse.ArgumentParser:
+    f.writelines(lines)
+
+
+def write_config_entry(*, entry_name: str, entry_value: Optional[str]) -> None:
```
Is there a reason this doesn't use [`configparser`](https://docs.python.org/3.7/library/configparser.html)?

cosmicexplorer reviewed (changes requested):
This could work, but if this is going to be the script that users run to setup pants, I would like some of the bash logic to be extracted into functions, as that also makes any weird errors more debuggable for the pants user. The `configparser` module seems like it might be able to abbreviate a lot of `ci.py`, but I'm less familiar with that.

Eric-Arellano committed:
Use configparser to dramatically simplify ci.py

Eric-Arellano committed:
Override ConfigParser's default of space around delimiter :

Eric-Arellano committed:
Refactor ./pants

Use functions and make it generally more robust!

Eric-Arellano committed:
Use sed line from Danny

Eric-Arellano committed:
Remove bad pants.ini line from debugging

cosmicexplorer commented on pants:
```diff
@@ -45,12 +44,41 @@ function tempdir {
+}
+
+function determine_pants_runtime_python_version {
+  interpreter_version="$(get_pants_ini_config_value pants_runtime_python_version)"
```
```suggestion
  interpreter_version="$(get_pants_ini_config_value 'pants_runtime_python_version')"
```
I tend to use the convention of surrounding string arguments with single quotes, especially when nested in a process output substitution, to differentiate them from function calls.

cosmicexplorer commented on pants:
```diff
@@ -45,12 +44,41 @@ function tempdir {
+  if [[ -z "${interpreter_version}" ]]; then
+    interpreter_version='2.7'
+  fi
+  python_bin_name=${PYTHON:-python${interpreter_version}}
```
```suggestion
  python_bin_name="${PYTHON:-python"${interpreter_version}"}"
```
Would definitely have double quotes around all variable dereferences, spaces can cause weird errors and consistency is key (imho). There's no reason not to have them on nested `${}` expressions too.

cosmicexplorer commented on pants:
```diff
@@ -45,12 +44,41 @@ function tempdir {
+  fi
+  python_bin_name=${PYTHON:-python${interpreter_version}}
+  if which "${python_bin_name}" >/dev/null; then
+    echo `which ${python_bin_name}`
```
Backtick substitution is "deprecated": http://mywiki.wooledge.org/BashFAQ/082. Many people cargo cult backticking from other scripts, especially for `which` calls, so you'll still see this often, but it should be converted to `"$()"`.

cosmicexplorer commented on pants:
```diff
@@ -45,12 +44,41 @@ function tempdir {
+  if which "${python_bin_name}" >/dev/null; then
+    echo `which ${python_bin_name}`
+  else
+    echo "${python_bin_name} could not be found on your PATH! Pants will not work." >&2
```
I don't super care about the distinction personally, but I would use `1>&2` here for consistency with the rest of the file.

cosmicexplorer commented on pants:
```diff
@@ -45,12 +44,41 @@ function tempdir {
 function bootstrap_venv {
-  if [[ ! -d "${PANTS_BOOTSTRAP}/${VENV_PACKAGE}" ]]
-  then
+  if [[ ! -d "${PANTS_BOOTSTRAP}/${VENV_PACKAGE}" ]]; then
```
Good change!

cosmicexplorer commented on pants.ini:
```diff
@@ -1,6 +1,6 @@
 plugins: [
-    'pantsbuild.pants.contrib.go==%(pants_version)s',
-  ]
+	'pantsbuild.pants.contrib.go==%(pants_version)s',
```
This changing is because of the use of `configparser`, right? This seems like an ok change then, since it seems like that module doesn't provide any options for spacing when writing out files.

cosmicexplorer commented on pants:
```diff
@@ -65,40 +93,40 @@ function bootstrap_venv {
 
-  if [[ ! -d "${PANTS_BOOTSTRAP}/${pants_version}" ]]
-  then
+  python_major_minor_version="$(get_python_major_minor_version ${python})"
```
```suggestion
  python_major_minor_version="$(get_python_major_minor_version "${python}")"
```
I would recommend doing this.

cosmicexplorer commented on pants:
```diff
@@ -65,40 +93,40 @@ function bootstrap_venv {
-  )
-  if [[ -n "${pants_version}" ]]
-  then
+  pants_version="$(get_pants_ini_config_value pants_version)"
```
```suggestion
  pants_version="$(get_pants_ini_config_value 'pants_version')"
```
I would recommend doing this.

cosmicexplorer committed:
Apply suggestions from Danny's code review

Co-Authored-By: Eric-Arellano <ericarellano@me.com>

Eric-Arellano committed:
Merge branch 'gh-pages' of github.com:pantsbuild/setup into py3

Eric-Arellano committed:
Additional bash improvements from Danny

Eric-Arellano committed:
Merge branch 'py3' of github.com:Eric-Arellano/setup into py3

Eric-Arellano committed:
Merge branch 'gh-pages' of github.com:pantsbuild/setup into py3

Eric-Arellano commented:
Reviewers: I broke out https://github.com/pantsbuild/setup/pull/36, https://github.com/pantsbuild/setup/pull/37, and https://github.com/pantsbuild/setup/pull/35 as pre-work for this PR, which are all merged now into this PR for a smaller diff.

This should be easier to review now and is all ready to go. I'm hoping to have it okayed before our release tomorrow, so that we can simply run the tests upon the release, prove it works via CI, and merge. Thanks!

Eric-Arellano committed:
Clarify in high level flow the env var name $PYTHON

We try to read in $PYTHON, which we store into $PYTHON_BIN_NAME. The important part is we check for the env var $PYTHON.

stuhood commented on pants:
```diff
@@ -62,8 +62,25 @@ function get_pants_ini_config_value {
+function check_interpreter_on_path {
+  python="$1"
+  if ! which "${python}" >/dev/null; then
+    die "Could not find ${python_bin_name}. Please ensure ${python_bin_name} is on your PATH."
```
It's implicit that `python_bin_name` will be set by this point... would probably be preferable to use the parameter or explicitly pass a human readable form if `$1` isn't readable.

Eric-Arellano replied:
This was 100% a bad merge! 

We intentionally want to keep this function generic, because in the future when we (maybe) auto-generate `pants.ini` I'm envisioning our flow will look something like this: `try py36 -> try py27 -> try py37 -> fail`.

This function—with modifications—will be the basis for that flow.

Thanks!

stuhood reviewed (approved):
Looks good: thanks!

Eric-Arellano committed:
Fix undefined name from bad merge

Thanks Stu!

Eric-Arellano committed:
Remove bad \ from .travis.yml script entry

Eric-Arellano committed:
Restore pants.ini if command fails and --python-version changed

Same fix we made in https://github.com/pantsbuild/setup/pull/36 for setting up pants versions. Forgot to also apply it to setup python version.

illicitonion commented on ci.py:
```diff
@@ -74,6 +93,25 @@ def setup_pants_version(test_pants_version: PantsVersion):
+  write_config(updated_config)
+  try:
+    yield
+  except subprocess.CalledProcessError:
```
I'm not sure why this `except` is here; isn't the behaviour the same without it?

illicitonion reviewed (approved):
Works for me, though I don't know really anything about the setup repo :)

Eric-Arellano committed:
Remove unnecessary except clause

Daniel pointed out we don't need this. We not only don't need it, but we shouldn't have it. No matter what, we need to do the cleanup, regardless of the exception caught.

Eric-Arellano committed:
Extract temporarily_rewrite_config() context manager

Do not repeat yourself. Makes it a bit more clear also why we need to read original config and have the try finally.

cosmicexplorer commented on pants:
```diff
@@ -72,11 +89,17 @@ function get_pants_ini_config_value {
+    python_bin_name="python${interpreter_version}"
   fi
+  check_interpreter_on_path "${python_bin_name}"
+  which "${python_bin_name}"
```
If you remove the `>/dev/null` in `check_interpreter_on_path` and rename it to `get_exe_path_or_die` (which I would recommend anyway, as it doesn't do anything python-specific), then you can remove this `which`. Checking that an exe exists with `which`, but then returning its path in another `which` call feels error-prone to me (what if we slightly change this method later and we end up checking that one interpreter exists, but returning a path to another one?). The suggestion isn't about code golfing -- checking that something exists and then return a path to it separately feels error-prone. I also find the result (just calling `get_exe_path_or_die "${python_bin_name}"` at the end, which will output the exe path to stdout on success) extremely readable -- I don't know that the `check_interpreter_on_path` function could potentially exit from looking at its callsites.

cosmicexplorer commented on ci.py:
```diff
@@ -22,9 +22,19 @@ def __str__(self):
       return self.value
 
 
+class PythonVersion(Enum):
```
Love enums!!

Eric-Arellano committed:
Use get_exe_path_or_die per Danny's suggestion

Eric-Arellano committed:
Revert bad setting of '-x'

Sorry for accidently pushing!

Eric-Arellano committed:
Rename local variables to reflect new function name

No longer hardcoded to Python.

Eric-Arellano committed:
Remove stray white line

cosmicexplorer reviewed (approved):
Thanks!!

Eric-Arellano committed:
Fix unnamed value in ci.py

We gotta get Pylint in this repo. Python only complains about unknown values once it's evaluated.

Eric-Arellano committed:
Upgrade to 1.15.0.dev4!

Eric-Arellano committed:
Upgrade virtualenv to 16.4.3

Running ./pants with Py3.6 or 3.7 would result in this error:
```
./pants
WARN] /Users/eric/.cache/pants/setup/bootstrap-Darwin-x86_64/1.15.0.dev4_py36/lib/python3.6/site.py:165: DeprecationWarning: 'U' mode is deprecated
  f = open(fullname, "rU")

WARN] /Users/eric/.cache/pants/setup/bootstrap-Darwin-x86_64/1.15.0.dev4_py36/lib/python3.6/site.py:165: DeprecationWarning: 'U' mode is deprecated
  f = open(fullname, "rU")

No goals specified.
Use `pants goals` to list goals.
Use `pants help` to get help.
```

To fix this, we must upgrade to 16.4.3, as the deprecation warning was fixed between 16.0.0 and this release.

Eric-Arellano committed:
See if python -m ensurepip works

I think the issue is that the wrong pip is being used? This looks like a Pip issue rather than twitter commons.

Eric-Arellano committed:
Revert "See if python -m ensurepip works"

This reverts commit fa1f2242e2ff3bdb872d41ea9c731babb696547d.

Still fails :/ but passes locally.

Eric-Arellano committed:
Squashed commit of the following:

commit 4799cfac1900fa8be3f5344142f407bd1896c657
Author: Eric Arellano <ericarellano@me.com>
Date:   Fri Mar 15 23:59:37 2019 -0700

    Remove pinned setuptools

Eric-Arellano committed:
Squashed commit of the following:

commit c7389044220edafdf6c8c0ca9b00c5eb960e2a16
Author: Eric Arellano <ericarellano@me.com>
Date:   Sat Mar 16 01:11:30 2019 -0700

    Remove bad leftover line

commit 4799cfac1900fa8be3f5344142f407bd1896c657
Author: Eric Arellano <ericarellano@me.com>
Date:   Fri Mar 15 23:59:37 2019 -0700

    Remove pinned setuptools

Eric-Arellano committed:
Merge branch 'gh-pages' of github.com:pantsbuild/setup into py3

Eric-Arellano committed:
Merge branch 'gh-pages' of github.com:pantsbuild/setup into py3

Eric-Arellano committed:
Fix bad merge leaving off --python-version arg

Eric-Arellano committed:
Remove bad __pycache__ from wip that made it into here
